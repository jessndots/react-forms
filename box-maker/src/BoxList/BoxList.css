.BoxList {
  display: flex;
  flex-flow: column wrap;
}

.BoxList-List {
  display: flex;
  flex-flow: row wrap;
}

.BoxList-Boxes {
  display: flex;
  flex-flow: row wrap;
  margin: 10px;
}

.BoxList-X {
  height: 50%
}