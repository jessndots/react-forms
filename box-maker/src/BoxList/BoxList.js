import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import Box from "../Box/Box.js";
import NewBoxForm from "../NewBoxForm/NewBoxForm.js"
import './BoxList.css';

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);

    const deleteBox = (id) => {
        const copy = [...boxes];
        setBoxes(copy.filter(box => box.id !== id))
    }

    const renderBoxes = () => {
        return (
            <ul className="BoxList-List">
                {boxes.map(box => (
                    <div className="BoxList-Boxes" key={box.id}>
                        <Box
                            backgroundColor={box.backgroundColor}
                            height={box.height}
                            width={box.width}
                            id={box.id}
                            deleteBox = {deleteBox}
                        />
                    </div>

                ))}
            </ul>
        )
    }

    /** Add new box object to list */
    const addBox = box => {
        let newBox = { ...box, id: uuid() };
        setBoxes(boxes => [...boxes, newBox]);
    }

    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
            {renderBoxes()}
        </div>
    )
}

export default BoxList;