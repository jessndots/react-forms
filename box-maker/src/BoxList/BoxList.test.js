import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import BoxList from './BoxList';


// Smoke Test
it("renders without crashing", function () {
  render(<BoxList />)
})

// Snapshot test
it("matches screenshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
})

it("adds new box", async function () {
  const { queryByText, getByText } = render(<BoxList />);

  // expect no boxes
  expect(queryByText("X")).not.toBeInTheDocument();

  // click Create Box
  fireEvent.click(getByText("Create Box"))

  // expect box
  await waitFor(() => {
    const removeBtn = getByText("X");
    expect(removeBtn).toBeInTheDocument();
    expect(removeBtn.nextElementSibling).toHaveStyle(`
      width: 50px;
      height: 50px;
  `)
  })
})

it("removes box", async function () {
  const {queryByText, getByText} = render(<BoxList/>);

  // create box
  fireEvent.click(getByText("Create Box"));

  // expect box
  await waitFor(() => {
    const removeBtn = getByText("X");
    expect(removeBtn).toBeInTheDocument();
    expect(removeBtn.nextElementSibling).toHaveStyle(`
      width: 50px;
      height: 50px;
  `)
  })

  // delete box
  fireEvent.click(getByText("X"));

  // expect no boxes
  expect(queryByText("X")).not.toBeInTheDocument();

})