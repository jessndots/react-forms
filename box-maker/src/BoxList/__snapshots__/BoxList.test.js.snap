// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches screenshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      action="#"
      class="NewBoxForm-Form"
    >
      <div>
        <label
          class="NewBoxForm-Label"
          for="backgroundColor"
        >
          Background Color
        </label>
        <input
          class="NewBoxForm-Field"
          name="backgroundColor"
          type="color"
          value="#000000"
        />
      </div>
      <div>
        <label
          for="height"
        >
          Height
        </label>
        <input
          class="NewBoxForm-Field"
          name="height"
          type="number"
          value="50"
        />
      </div>
      <div>
        <label
          for="width"
        >
          Width
        </label>
        <input
          class="NewBoxForm-Field"
          name="width"
          type="number"
          value="50"
        />
      </div>
      <div>
        <button
          type="submit"
        >
          Create Box
        </button>
      </div>
    </form>
    <ul
      class="BoxList-List"
    />
  </div>
</DocumentFragment>
`;
